core/cont/inc/TCollection.h

```diff
@@ -272,6 +272,60 @@ inline TIter TCollection::begin() const { return ++(TIter(this)); }
 inline TIter TCollection::end() const { return TIter::End(); }
 
 
+
+/// @brief TTypedIter is a typed version of TIter.
+///
+/// The typical use is:
+/// ~~~ {.cpp}
+///    TTypedIter<TBaseClass> next(cl->GetListOfBases());
+///    while(auto bcl = next()) {
+///       ... use bcl as a TBaseClass*
+///    }
+/// ~~~ {.cpp}
+template <class Containee> // Containee must derive from TObject.
+class TTypedIter : public TIter
+{
+   static_assert(std::is_base_of<TObject, Containee>::value,
+                 "Containee type must inherit from TObject");
+
+   /// This is a workaround against ClassDefInline not supporting classes
+   /// missing their default constructor or having them private.
+   template <class T> friend class ROOT::Internal::ClassDefGenerateInitInstanceLocalInjector;
+
+   TTypedIter() = default;
+
+public:
+
+   using TIter::TIter;
+   TTypedIter(const TIter &iter) : TIter(iter) {}
+
+   Containee *operator()() { return static_cast<Containee *>(TIter::Next()); }
```

Comment:
This has proven (to me and users) to be very dangerous. Many of ROOT's lists are heterogenous, even though people might not realize and the collection getter names might suggest otherwise - e.g. `GetListOfStreamerInfos()`. I'd really prefer this and the others to be dynamic casts. Yes more costly in CPU time, but the debugging cost of user code that this will otherwise cause is non-negligible.

Reply:
However, you can not have filtering/dynamic-cast in a TIter... as the usage is:
```
while( .... = iterator.operator() ) ....
```
which means that any unexpected element would silently trim the collection's iteration.  For TRangeCast on the other hand it might help.

Reply:
Then `TIter` should really throw upon "invalid" `dynamic_cast`